### PR TITLE
feat: add breakpoint column support

### LINKED ISSUE #83: Multiple breakpoints on the same line
var list = new List<int>() {1,2,3,4,5,6,7,8,9};
       var lst1 = list.Where(i => i > 1). Where(i => i< 6).ToList();
                                   ^                  ^
                                  bp1                bp2   
Set two breakpoint at same line.

Breakpoint 2 will not be hitted when breakpoint 1 is htted and continue.
Can breakpoints_line support columns?

### PR DESCRIPTION
**Pull request by GustavEikaas**
My attempt at implementing what was discussed in #219 and resolves #83.                                                                                                                 
                                                                                                                                                                                          
As discussed with @viewizard, the approach is: use the column to pick the right IL offset while keeping the 1 breakpoint per line rule. Column 0 falls back to existing behavior.                                                                                                                                                                              

I had to make some heuristics changes in SymbolReader.cs, feel free to sanity check me. 
    
When a column is provided, sequence point selection for the outer method prefers the SP whose StartColumn is closest to (but not exceeding) the target column. A cursorInsideNested guard ensures that when the cursor is positioned inside a lambda's column range, the breakpoint lands in the lambda rather than the outer call site.

column and endColumn are now returned in the DAP setBreakpoints response.                                                                                                               
 
The test case from #83 now works. Tested locally in neovim with various changes and seems to work as expected

Keep in mind I usually never code in C++. If there are edge cases I've missed feel free to tag me and I'll test and fix.

Fixes #83

### CODE REVIEW

**Comment by viewizard** [comment]
This is a large change that should be covered by tests. Please add tests to VSCodeTestSrcBreakpointResolve. You will probably need to add a `Context.AddBreakpoint()` implementation that has a third operand (column) that should be set manually (I'm not really sure how we could add the same idea as for lines with `Label.Breakpoint()` for auto line number calculation).

Please cover all possible cases (proper resolve with column inside statement, several breakpoints set to the same statement, etc.), for example:

1) Multiline code, set breakpoint with column on a non-first line. Breakpoint should be set at the start of the statement.
```
int a = 
a + 25 +                  <<<=== breakpoint on this line with column 3, for example
ff + 5;
```

2) Multiline code with code after it:
```
int a = 
a + 25 +                  
ff + 5; int a = 5;                 <<<=== breakpoint on this line with column 15 (int a = 5), for example
```

3) Multiline code:
```
int a = 
a + 25 +                            <<<=== first breakpoint on this line with column 3, for example
ff + 5; int a = 5;                 <<<=== second breakpoint on this line with column 15 (int a = 5), for example
```

4) Multiline code with code after it, with search for proper next statement:
```
int a = 
a + 25 +                  
ff + 5; int a = 5;                 <<<=== breakpoint on this line with column 55 (out of int a = 5), for example
int bbb = 10;                     <<<=== real resolved breakpoint location
```

5) Multiple breakpoints set for one line:
```
int i = 0; i++; i--; i = 5;
```
First set with column 0, then with 17, then with 23.

6) Multiple breakpoints set for one line:
```
int i = 0; i++; i--; i = 5;
```
First set with column 17, then with 0, then with 23.

And so on...

**Comment by GustavEikaas** [comment]
Yep I agreee, ill make some tests. I did try to run the tests because I considered adding some but ran into some issues with dotnet runtime 3.1 and globalization issues. Ill figure that out and cover it with tests

**Comment by viewizard** [comment]
You could replace "netcoreapp3.1" with "net10.0" in all ./test-suite files and use .NET 10 as I do ;)

**Comment by GustavEikaas** [comment]
Ah lol didnt think of that, is it worth considering upstreaming that change or does it break CI stuff?

**Comment by viewizard** [comment]
CI tests with .NET 3.1, 6.0, 8.0 and 10.0 now in the same way as I wrote - it replace "netcoreapp3.1" with the framework it needs for the current test loop. :)

**Comment by GustavEikaas** [comment]
Sorry it took a while, I added some tests now. let me know what you think

**Comment by GustavEikaas** [comment]
@viewizard **nudge**

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs [hunk 1]
```diff
@@ -209,15 +214,113 @@ public void WasBreakpointHit(string caller_trace, string bpName)
...
+            var lbp = (LineBreakpoint)breakpoint;
+
+            StackTraceResponse stackTraceResponse =
+                JsonConvert.DeserializeObject<StackTraceResponse>(ret.ResponseStr);
+
+            if (stackTraceResponse.body.stackFrames[0].line == lbp.NumLine
+                && stackTraceResponse.body.stackFrames[0].source.name == lbp.FileName
+                && stackTraceResponse.body.stackFrames[0].column == expectedColumn
+                && stackTraceResponse.body.stackFrames[0].endColumn == expectedEndColumn)
+                return;
```
Comment (viewizard) [inline test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: You could probably use Assert.Equal() directly here with the response and expected values, IMHO — there's no reason to check values and then throw an exception in case of failure.

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs [hunk 2]
```diff
@@ -209,15 +214,113 @@ public void WasBreakpointHit(string caller_trace, string bpName)
...
+            var ret = VSCodeDebugger.Request(stackTraceRequest);
+            Assert.True(ret.Success, @"__FILE__:__LINE__"+"\n"+caller_trace);
+
+            Breakpoint breakpoint = ControlInfo.Breakpoints[bpName];
+            var lbp = (LineBreakpoint)breakpoint;
+            StackTraceResponse stackTraceResponse = JsonConvert.DeserializeObject<StackTraceResponse>(ret.ResponseStr);
+
+            if (stackTraceResponse.body.stackFrames[0].line == lbp.NumLine
+                && stackTraceResponse.body.stackFrames[0].source.name == lbp.FileName
+                && stackTraceResponse.body.stackFrames[0].column == expectedColumn)
```
Comment (viewizard) [inline test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: You could probably use Assert.Equal() directly here with the response and expected values, IMHO — there's no reason to check values and then throw an exception in case of failure.

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs [hunk 3]
```diff
@@ -209,15 +214,113 @@ public void WasBreakpointHit(string caller_trace, string bpName)
...
+            var ret = VSCodeDebugger.Request(stackTraceRequest);
+            Assert.True(ret.Success, @"__FILE__:__LINE__"+"\n"+caller_trace);
+
+            StackTraceResponse stackTraceResponse =
+                JsonConvert.DeserializeObject<StackTraceResponse>(ret.ResponseStr);
+
+            if (stackTraceResponse.body.stackFrames[0].line == bp_line
+                && stackTraceResponse.body.stackFrames[0].source.name == bp_fileName
+                && stackTraceResponse.body.stackFrames[0].column == expectedColumn)
+                return;
```
Comment (viewizard) [inline test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: You could probably use Assert.Equal() directly here with the response and expected values, IMHO — there's no reason to check values and then throw an exception in case of failure.

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs [hunk 4]
```diff
@@ -595,11 +698,70 @@ void nested_func11() { void nested_func12() { void nested_func13() {
...
+                Context.AddBreakpoint(@"__FILE__:__LINE__", "bp_col3", Column: 43);
+                Context.AddBreakpoint(@"__FILE__:__LINE__", "bp_col4", Column: 35);
+
+                int endLine = Context.GetBreakpointLine("multi_end");
+
+                // Multiline statement: breakpoint set with a column on the non-first line snaps up to the start of the statement.
+                Context.AddManualBreakpoint(@"__FILE__:__LINE__", "Program.cs", endLine - 1, Column: 5);
+
+                // Multiple breakpoints on the same line coalesce to a single BP at the start of the line, max 1 breakpoint-per line rule
+                Context.AddBreakpoint(@"__FILE__:__LINE__", "bp_multi_order_A", Column: 0);
```
Comment (viewizard) [inline test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: This one looks strange — add a comment about disabling column and line-only breakpoint setup, or better yet, remove 0 altogether and make it the default value for the Column.

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs (hunk 1 above)
Comment (GustavEikaas) [reply test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: fixed

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs (hunk 2 above)
Comment (GustavEikaas) [reply test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: fixed

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs (hunk 3 above)
Comment (GustavEikaas) [reply test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: fixed

File: test-suite/VSCodeTestSrcBreakpointResolve/Program.cs (hunk 4 above)
Comment (GustavEikaas) [reply test-suite/VSCodeTestSrcBreakpointResolve/Program.cs]: fixed

**Comment by viewizard** [comment]
Looks good. I'm fixing the Samsung CI now. As soon as it's up, we can start testing these changes.
